cni/network/network.go

```diff
@@ -457,199 +454,243 @@ func (plugin *netPlugin) Add(args *cniSkel.CmdArgs) error {
 			return err
 		}
 
-		nwDNSInfo, err = getNetworkDNSSettings(nwCfg, result, k8sNamespace)
-		if err != nil {
-			err = plugin.Errorf("Failed to getDNSSettings: %v", err)
-			return err
-		}
+		defer func() {
+			err = plugin.cleanupAllocationOnError(err, result, resultV6, nwCfg, args, options)
+		}()
+	}
 
-		log.Printf("[cni-net] nwDNSInfo: %v", nwDNSInfo)
-		// Update subnet prefix for multi-tenant scenario
-		if err = updateSubnetPrefix(cnsNetworkConfig, &subnetPrefix); err != nil {
-			err = plugin.Errorf("Failed to updateSubnetPrefix: %v", err)
+	// Create network
+	if nwInfoErr != nil {
+		// Network does not exist.
+		log.Printf("[cni-net] Creating network %v.", networkID)
+		if nwInfo, err = plugin.createNetworkInternal(networkID, policies, args, nwCfg, cnsNetworkConfig, subnetPrefix, result, resultV6); err != nil {
+			log.Errorf("Create network failed:%w", err)
 			return err
 		}
 
-		// Create the network.
-		nwInfo = network.NetworkInfo{
-			Id:           networkId,
-			Mode:         nwCfg.Mode,
-			MasterIfName: masterIfName,
-			AdapterName:  nwCfg.AdapterName,
-			Subnets: []network.SubnetInfo{
-				{
-					Family:  platform.AfINET,
-					Prefix:  subnetPrefix,
-					Gateway: gateway,
-				},
-			},
-			BridgeName:                    nwCfg.Bridge,
-			EnableSnatOnHost:              nwCfg.EnableSnatOnHost,
-			DNS:                           nwDNSInfo,
-			Policies:                      policies,
-			NetNs:                         args.Netns,
-			DisableHairpinOnHostInterface: nwCfg.DisableHairpinOnHostInterface,
-			IPV6Mode:                      nwCfg.IPV6Mode,
-			ServiceCidrs:                  nwCfg.ServiceCidrs,
-		}
+		log.Printf("[cni-net] Created network %v with subnet %v.", networkID, subnetPrefix.String())
+	}
 
-		nwInfo.IPAMType = nwCfg.Ipam.Type
+	epInfo, err := plugin.createEndpointInternal(nwCfg, cnsNetworkConfig, result, resultV6, azIpamResult, args, &nwInfo,
+		policies, endpointId, k8sPodName, k8sNamespace, enableInfraVnet, enableSnatForDns)
+	if err != nil {
+		log.Errorf("Endpoint creation failed:%w", err)
+		return err
+	}
 
-		if len(result.IPs) > 0 {
-			_, podnetwork, err := net.ParseCIDR(result.IPs[0].Address.String())
-			if err != nil {
-				return err
-			}
+	msg := fmt.Sprintf("CNI ADD succeeded : CNI Version %+v, IP:%+v, VlanID: %v, Interfaces:%+v, podname %v, namespace %v",
+		result.CNIVersion, result.IPs, epInfo.Data[network.VlanIDKey], result.Interfaces, k8sPodName, k8sNamespace)
+	plugin.setCNIReportDetails(nwCfg, CNI_ADD, msg)
+
+	return nil
+}
 
-			nwInfo.PodSubnet = network.SubnetInfo{
-				Family:  platform.GetAddressFamily(&result.IPs[0].Address.IP),
-				Prefix:  *podnetwork,
-				Gateway: result.IPs[0].Gateway,
+func (plugin *NetPlugin) cleanupAllocationOnError(err error,
+	result, resultV6 *cniTypesCurr.Result,
+	nwCfg *cni.NetworkConfig,
+	args *cniSkel.CmdArgs,
+	options map[string]interface{}) error {
+	if err != nil {
```

Comment:
method name cleanupAllocationOnError is descript enough that client can invoke on error instead of passing and checking here